### DISCUSSION TITLE
[Suggestion] Possibility to change the order of displaying messages in the admin area

### DISCUSSION THREAD

**Discussion by nsklyarov** [rabbitmq/rabbitmq-server#12718] [Ideas]
### RabbitMQ series

3.13.x

### Operating system (distribution) used

macOS Sonoma 14.05

### How is RabbitMQ deployed?

Community Docker image

### What would you like to suggest for a future version of RabbitMQ?

Hello RabbitMQ team!

I would like to discuss the addition of a new UI feature in the RabbitMQ admin - the ability to change the order in which the message list is displayed (from oldest to newest and vice versa). 

This feature, in my opinion, can greatly improve the UX when dealing with large queues of messages.

<img width="1224" alt="RabbitMQReverseOrder" src="https://github.com/user-attachments/assets/d16e9131-57d6-4d06-afb1-642bce464948" />

### Context
In the current interface, messages can only be viewed in one order. 
When there are many messages, being able to see them in reverse order allows you to find the last or first messages in the queue faster, which can be useful for debugging.

### Suggestion
Add an interface option to the ‘Get Messages’ section, such as a ‘Reverse Order’ checkbox, that would allow users to choose the order in which messages are displayed.

### Related PR
I have previously attempted to implement this feature in PR [#12698](https://github.com/rabbitmq/rabbitmq-server/pull/12698), but have reworked the approach to suit current requirements and constraints in these PR [#12717](https://github.com/rabbitmq/rabbitmq-server/pull/12717).

I would be happy to hear your thoughts and discuss how this idea could be improved or adapted for adoption into the project.
Thanks for your attention and feedback!

**Comment by michaelklishin** [comment]
I don't want to touch that feature or `basic.get` until we get a `rabbitmq-queues peek`-like option to get N messages and get rid of `basic.get` everywhere except for integration tests. And it's simply not a priority or commonly requested feature.

Except for certain RabbitMQ-as-a-Service environments, this arguably should simply be a `rabbitmq-queues peek` extension of sorts.

But I've asked the rest of the core team to chime in.

**Comment by michaelklishin** [comment]
Besides the fact that every contribution the core team accepts will be extra weight on our shoulders 6, 12, 24, 36 months from now.

And the fact that those who need a customized management UI can build it using the same HTTP API the management plugin uses. It can be entirely external to RabbitMQ, not have to use the JS code fundamentally from over a decade ago, and have to run their ideas by our team. Complete freedom to scratch your own itch with a moderate amount of effort (you don't need to reimplement the entire management UI).

**Comment by deadtrickster** [comment]
If a queue is short to find the latest message you just need to scroll a bit. If it long you have to build a long long list with basic.get. Which means if this checkbook set on a production system Operator will get a surprise. From PR I can't conclude this checkbox is shown only for short queues or has production safeguards.

**Reply by nsklyarov** [reply]
Excuse me, I can't agree.
With flag "Reverse" or without - complexity O(n)...

Could you offer me a better API or part of the documentation that I paid attention to so that I could write really long-term code.
Well, I can use the plugin abstraction, but it seems that it's too simple a thing to create a plugin.

**Reply by deadtrickster** [reply]
you write "to find the last or first messages". To get the last message of the queue you have to go thru the whole queue potentially invalidating all the caches and other strange stuff. Plus I don't even understand what would be the last message on a running system anyway.

**Comment by kjnilsson** [comment]
If your queue is of type "quorum" you can use the command line utility to peek at a message at a given position:

`rabbitmq-queues [--node <node>] [--longnames] [--quiet] peek [--vhost <vhost>] <queue> <position> [--timeout <timeout>]`
